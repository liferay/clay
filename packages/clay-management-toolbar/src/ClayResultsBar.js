import 'clay-button';
import 'clay-label';
import 'clay-link';
import {Config} from 'metal-state';
import ClayComponent from 'clay-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';

import {filterLabelsValidator} from './validators';

import templates from './ClayResultsBar.soy.js';

/**
 * Metal ClayResultsBar component.
 * @extends ClayComponent
 */
class ClayResultsBar extends ClayComponent {
	/**
	 * Continues the propagation of the clear button clicked event
	 * @param {!Event} event
	 * @private
	 */
	_handleClearResultsClick(event) {
		this.emit('clearButtonClicked', event);
	}

	/**
	 * Continues the propagation of the filter label close clicked event
	 * @param {!Event} event
	 * @private
	 */
	_handleFilterLabelCloseClicked(event) {
		this.emit('filterLabelCloseClicked', event);
	}
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayResultsBar.STATE = {
	/**
	 * Url for clear results link.
	 * @default undefined
	 * @instance
	 * @memberof ClayResultsBar
	 * @type {?(string|undefined)}
	 */
	clearResultsURL: Config.string(),

	/**
	 * List of filter label items.
	 * @default undefined
	 * @instance
	 * @memberof ClayResultsBar
	 * @type {?(array|undefined)}
	 */
	filterLabels: filterLabelsValidator.required(),

	/**
	 * Value of the search input.
	 * @default undefined
	 * @instance
	 * @memberof ClayResultsBar
	 * @type {?(string|undefined)}
	 */
	searchValue: Config.string(),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @default undefined
	 * @instance
	 * @memberof ClayResultsBar
	 * @type {?(string|undefined)}
	 */
	spritemap: Config.string().required(),

	/**
	 * Total number of items. If totalItems is 0 most of the elements in the bar
	 * will appear disabled.
	 * @default undefined
	 * @instance
	 * @memberof ClayResultsBar
	 * @type {!number}
	 */
	totalItems: Config.number().required(),
};

defineWebComponent('clay-results-bar', ClayResultsBar);

Soy.register(ClayResultsBar, templates);

export {ClayResultsBar};
export default ClayResultsBar;
