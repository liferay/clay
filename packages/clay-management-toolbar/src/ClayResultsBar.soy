{namespace ClayResultsBar}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param filterLabels: list<?>}
	{@param spritemap: string}
	{@param totalItems: number}
	{@param? _handleClearResultsClick: any}
	{@param? _handleFilterLabelCloseClicked: any}
	{@param? clearResultsURL: string}
	{@param? searchValue: string}

	<nav class="tbar tbar-inline-xs-down subnav-tbar subnav-tbar-primary">
		<div class="container-fluid container-fluid-max-xl">
			<ul class="tbar-nav tbar-nav-wrap">
				{let $hasFilterLabels: length($filterLabels) > 0 /}

				{call .results}
					{param expand: not $hasFilterLabels /}
					{param searchValue: $searchValue /}
					{param totalItems: $totalItems /}
				{/call}

				{if $hasFilterLabels}
					{call .filterLabels}
						{param _handleFilterLabelCloseClicked: $_handleFilterLabelCloseClicked /}	
						{param filterLabels: $filterLabels /}	
						{param spritemap: $spritemap /}	
					{/call}
				{/if}

				{call .clearResults}
					{param _handleClearResultsClick: $_handleClearResultsClick /}
					{param clearResultsURL: $clearResultsURL /}
				{/call}
			</ul>
		</div>
	</nav>
{/template}

/**
 * This renders de clear results button or link
 */
{template .clearResults}
	{@param? _handleClearResultsClick: any}
	{@param? clearResultsURL: string}

	<li class="tbar-item">
		<div class="tbar-section">
			{if $clearResultsURL}
				{call ClayLink.render}
					{param elementClasses: 'component-link tbar-link' /}
					{param events: ['click': $_handleClearResultsClick] /}
					{param href: $clearResultsURL /}
					{param label kind="text"}
						{msg desc="Label for clear results link"}
							clear
						{/msg}
					{/param}

					{param ref: 'clearSearch' /}
				{/call}
			{else}
				{call ClayButton.render}
					{param elementClasses: 'component-link tbar-link' /}
					{param events: ['click': $_handleClearResultsClick] /}
					{param label kind="text"}
						{msg desc="Label for clear results button"}
							clear
						{/msg}
					{/param}

					{param ref: 'clearSearch' /}
					{param style: 'unstyled' /}
				{/call}
			{/if}
		</div>
	</li>
{/template}

/**
 * This renders the list of filter labels
 */
{template .filterLabels}
	{@param filterLabels: list<?>}
	{@param spritemap: string}
	{@param? _handleFilterLabelCloseClicked: any}

	{foreach $filterLabel in $filterLabels}
		{let $clayFilterLabelClasses kind="text"}
			tbar-item
			{if isLast($filterLabel)}
				{sp}tbar-item-expand
			{/if}
		{/let}

		<li class="{$clayFilterLabelClasses}">
			<div class="tbar-section">
				{call ClayFilterLabel.render}
					{param closeable: $filterLabel.closeable ?: true /}
					{param data: ['labelId': index($filterLabel)] /}
					{param events: ['filterLabelCloseClicked': $_handleFilterLabelCloseClicked ]/}
					{param filterData: $filterLabel.data /}
					{param label: $filterLabel.label /}
					{param ref: 'label' + index($filterLabel) /}
					{param spritemap: $spritemap /}
				{/call}
			</div>
		</li>
	{/foreach}
{/template}

/**
 * This renders the text indicatin how many results there are
 */
{template .results}
	{@param totalItems: number}
	{@param? expand: bool}
	{@param? searchValue: string}

	{let $elementClasses kind="text"}
		tbar-item
		{if $expand}
			{sp}tbar-item-expand
		{/if}
	{/let}

	<li class="{$elementClasses}">
		<div class="tbar-section">
			<span class="component-text text-truncate-inline">
				<span class="text-truncate">
					{if $searchValue}
						{msg desc="Says how many results there are for a search"}
							{$totalItems}-results-for-{$searchValue}
						{/msg}
					{else}
						{msg desc="Says how many results there are"}
							{$totalItems}-results-for
						{/msg}
					{/if}
				</span>
			</span>
		</div>
	</li>
{/template}