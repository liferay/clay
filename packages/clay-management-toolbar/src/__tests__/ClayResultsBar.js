import ClayResultsBar from '../ClayResultsBar';

let resultsBar;

const filterLabels = [
	{
		closeable: false,
		label: 'Label 1',
	},
	{
		label: 'Label 2',
	},
	{
		label: 'Label 3',
	},
];

const spritemap = '../node_modules/clay-css/lib/images/icons/icons.svg';

describe('ClayResultsBar', function() {
	afterEach(() => {
		if (resultsBar) {
			resultsBar.dispose();
		}
	});

	it('should render the default markup', () => {
		resultsBar = new ClayResultsBar({
			spritemap: spritemap,
			totalItems: 10,
		});

		expect(resultsBar).toMatchSnapshot();
	});

	it('should render a results bar with search value', () => {
		resultsBar = new ClayResultsBar({
			searchValue: 'my search',
			spritemap: spritemap,
			totalItems: 10,
		});

		expect(resultsBar).toMatchSnapshot();
	});

	it('should render a results bar with filter labels', () => {
		resultsBar = new ClayResultsBar({
			filterLabels: filterLabels,
			spritemap: spritemap,
			totalItems: 10,
		});

		expect(resultsBar).toMatchSnapshot();
	});

	it('should render a results bar with filter labels and search value', () => {
		resultsBar = new ClayResultsBar({
			filterLabels: filterLabels,
			searchValue: 'my search',
			spritemap: spritemap,
			totalItems: 10,
		});

		expect(resultsBar).toMatchSnapshot();
	});

	it('should render a results bar and emit an event on clear button click', () => {
		resultsBar = new ClayResultsBar({
			searchValue: 'my search',
			spritemap: spritemap,
			totalItems: 10,
		});

		const spy = jest.spyOn(resultsBar, 'emit');

		resultsBar.refs.clearSearch.element.click();

		expect(spy).toHaveBeenCalled();
		expect(spy).toHaveBeenCalledWith(
			'clearButtonClicked',
			expect.any(Object)
		);
	});

	it('should render a results bar and emit an event on close label click', () => {
		resultsBar = new ClayResultsBar({
			filterLabels: filterLabels,
			spritemap: spritemap,
			totalItems: 10,
		});

		const spy = jest.spyOn(resultsBar, 'emit');

		resultsBar.refs.label1.refs.closeButton.element.click();

		expect(spy).toHaveBeenCalled();
		expect(spy).toHaveBeenCalledWith(
			'filterLabelCloseClicked',
			expect.any(Object)
		);
	});
});
