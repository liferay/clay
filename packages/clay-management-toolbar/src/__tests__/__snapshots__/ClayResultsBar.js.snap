// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayResultsBar should render a results bar with filter labels 1`] = `
<nav class="tbar tbar-inline-md-down subnav-tbar subnav-tbar-primary">
  <div class="container-fluid container-fluid-max-xl">
    <ul class="tbar-nav tbar-nav-wrap">
      <li class="tbar-item">
        <div class="tbar-section">
          <span class="component-text text-truncate-inline">
            <span class="text-truncate">10-results-for</span>
          </span>
        </div>
      </li>
      <li class="tbar-item">
        <div class="tbar-section">
          <span class="label component-label tbar-label">
            <span class="label-item label-item-expand">Label 1</span>
          </span>
        </div>
      </li>
      <li class="tbar-item">
        <div class="tbar-section">
          <span class="label label-dismissible component-label tbar-label">
            <span class="label-item label-item-expand">Label 2</span>
            <span class="label-item label-item-after">
              <button class="btn close" aria-label="close" type="button">
                <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times" focusable="false">
                  <title>times</title>
                  <use xlink:href="../node_modules/clay-css/lib/images/icons/icons.svg#times"></use>
                </svg>
              </button>
            </span>
          </span>
        </div>
      </li>
      <li class="tbar-item tbar-item-expand">
        <div class="tbar-section">
          <span class="label label-dismissible component-label tbar-label">
            <span class="label-item label-item-expand">
              <strong>Category:</strong>Label 3</span>
            <span class="label-item label-item-after">
              <button class="btn close" aria-label="close" type="button">
                <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times" focusable="false">
                  <title>times</title>
                  <use xlink:href="../node_modules/clay-css/lib/images/icons/icons.svg#times"></use>
                </svg>
              </button>
            </span>
          </span>
        </div>
      </li>
      <li class="tbar-item">
        <button class="btn component-link tbar-link btn-unstyled" type="button">clear</button>
      </li>
    </ul>
  </div>
</nav>
`;

exports[`ClayResultsBar should render a results bar with filter labels and search value 1`] = `
<nav class="tbar tbar-inline-md-down subnav-tbar subnav-tbar-primary">
  <div class="container-fluid container-fluid-max-xl">
    <ul class="tbar-nav tbar-nav-wrap">
      <li class="tbar-item">
        <div class="tbar-section">
          <span class="component-text text-truncate-inline">
            <span class="text-truncate">10-results-for-my search</span>
          </span>
        </div>
      </li>
      <li class="tbar-item">
        <div class="tbar-section">
          <span class="label component-label tbar-label">
            <span class="label-item label-item-expand">Label 1</span>
          </span>
        </div>
      </li>
      <li class="tbar-item">
        <div class="tbar-section">
          <span class="label label-dismissible component-label tbar-label">
            <span class="label-item label-item-expand">Label 2</span>
            <span class="label-item label-item-after">
              <button class="btn close" aria-label="close" type="button">
                <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times" focusable="false">
                  <title>times</title>
                  <use xlink:href="../node_modules/clay-css/lib/images/icons/icons.svg#times"></use>
                </svg>
              </button>
            </span>
          </span>
        </div>
      </li>
      <li class="tbar-item tbar-item-expand">
        <div class="tbar-section">
          <span class="label label-dismissible component-label tbar-label">
            <span class="label-item label-item-expand">
              <strong>Category:</strong>Label 3</span>
            <span class="label-item label-item-after">
              <button class="btn close" aria-label="close" type="button">
                <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times" focusable="false">
                  <title>times</title>
                  <use xlink:href="../node_modules/clay-css/lib/images/icons/icons.svg#times"></use>
                </svg>
              </button>
            </span>
          </span>
        </div>
      </li>
      <li class="tbar-item">
        <button class="btn component-link tbar-link btn-unstyled" type="button">clear</button>
      </li>
    </ul>
  </div>
</nav>
`;

exports[`ClayResultsBar should render a results bar with search value 1`] = `
<nav class="tbar tbar-inline-md-down subnav-tbar subnav-tbar-primary">
  <div class="container-fluid container-fluid-max-xl">
    <ul class="tbar-nav tbar-nav-wrap">
      <li class="tbar-item tbar-item-expand">
        <div class="tbar-section">
          <span class="component-text text-truncate-inline">
            <span class="text-truncate">10-results-for-my search</span>
          </span>
        </div>
      </li>
      <li class="tbar-item">
        <button class="btn component-link tbar-link btn-unstyled" type="button">clear</button>
      </li>
    </ul>
  </div>
</nav>
`;

exports[`ClayResultsBar should render the default markup 1`] = `
<nav class="tbar tbar-inline-md-down subnav-tbar subnav-tbar-primary">
  <div class="container-fluid container-fluid-max-xl">
    <ul class="tbar-nav tbar-nav-wrap">
      <li class="tbar-item tbar-item-expand">
        <div class="tbar-section">
          <span class="component-text text-truncate-inline">
            <span class="text-truncate">10-results-for</span>
          </span>
        </div>
      </li>
      <li class="tbar-item">
        <button class="btn component-link tbar-link btn-unstyled" type="button">clear</button>
      </li>
    </ul>
  </div>
</nav>
`;
